 **Strings in Python**
======================

In Python, a string is a sequence of characters enclosed in quotes (either single, double, or triple quotes). Strings are `immutable`, meaning they cannot be changed after they are created.

**Creating Strings**
-------------------

There are several ways to create strings in Python:

1. **Single Quotes**: `my_string = 'Hello, World!'`
2. **Double Quotes**: `my_string = "Hello, World!"`
3. **Triple Quotes**: `my_string = '''Hello, World!'''` (can span multiple lines)
4. **Raw Strings**: `my_string = r'Hello, World!'` (treats backslashes as literal characters)

In [ ]:
#for multi line string use triple quotes '''any string'''
my_string: str = '''Hello,
World!'''
print(my_string)

Hello,
World!


In [ ]:
my_string: str = r'Hello\t,\n Worl\\d!'
print(my_string)

print("\n-----\n")

my_string = 'Hello,\n World!'
print(my_string)

Hello\t,\n Worl\\d!

-----

Hello,
 World!


**Escape Sequence Characters in Python**
======================================

In Python, escape sequence characters are used to represent special characters that have a specific meaning in a string. These characters are denoted by a backslash (`\`) followed by a character.


In [ ]:
print("Hello,\b World!") #\b backspace
print("Hello,\tWorld!")  #\t tab
print("Hello, \"World!\"")
print("Hello,\\ World!")

Hello, World!
Hello,	World!
Hello, "World!"
Hello,\ World!


### **Unicode characters**


Python also supports Unicode escape sequence characters, which are used to represent `Unicode characters`. These characters are denoted by `\u` followed by a four-digit hexadecimal code.

In [ ]:
print(r"\u0041 = ", "\u0041")
print(r"\u0042 = ", "\u0042")
print(r"\u0043 = ", "\u0043")

\u0041 =  A
\u0042 =  B
\u0043 =  C


### **Performing Differrent Operations on String Object**




| S# | Operation         | Example                                                                           |
|----|-------------------|-----------------------------------------------------------------------------------|
| 1  | **Concatenation** | `my_string = 'Hello, ' + 'World!'`                                                |
| 2  | **Indexing**      | `my_string = 'Hello, World!'; print(my_string[0])  # prints 'H'`                  |
| 3  | **Slicing**       | `my_string = 'Hello, World!'; print(my_string[7:])  # prints 'World!'`            |
| 4  | **Length**        | `my_string = 'Hello, World!'; print(len(my_string))  # prints 13`                 |
| 5  |  **Upper Case**   | `my_string = 'Hello, World!'; print(my_string.upper())  # prints 'HELLO, WORLD!'` |
| 6  | **Lower Case**    | `my_string = 'Hello, World!'; print(my_string.lower())  # prints 'hello, world!'` |

In [ ]:
my_string = 'Hello, ' + 'World!' #Concatenation using + sign
print(my_string)

Hello, World!


In [ ]:
#Indexing, index value starts with 0 zero, so  the first character
#have index vlaue 0, second character have index value 1 and the third one
#have index value 2 and so on.
print(my_string[1]) #It will print 'e'

e


In [ ]:
my_string: str = 'Hello, World!'
print(my_string[7:]) #It starts from 7 till the end of the string
print(my_string[0:5]) #It starts from 0 till the index 4 - (0,1,2,3,4) = 5 characters

World!
Hello


In [ ]:
print(len(" Hello, World! "))#calculating length of a string even the space will be treated as character

15


In [ ]:
print(my_string.upper()) #Upper Case
print(my_string.lower()) #Lower Case

HELLO, WORLD!
hello, world!


Here are some commonly used string methods:

1. **`split()`**: splits a string into a list of substrings based on a delimiter
2. **`join()`**: joins a list of strings into a single string
3. **`replace()`**: replaces a substring with another substring
4. **`find()`**: returns the index of a substring
5. **`count()`**: returns the number of occurrences of a substring

In [2]:
my_string: str = 'Hello! World'

# split into a list of words
words: str = my_string.split()
print("my_string.split()    = ", words)

words = my_string.split(" ") # Space as a delimiter
print('my_string.split(" ") = ',words)

words = my_string.split("l") # Splitting using 'l' as the delimiter
print('my_string.split("l") = ', words)

my_string.split()    =  ['Hello!', 'World']
my_string.split(" ") =  ['Hello!', 'World']
my_string.split("l") =  ['He', '', 'o! Wor', 'd']


**Steps of Execution:**

- The split("l") method breaks the string "Hello! World" wherever the letter l appears.

<br>

After splitting, the parts are:

<br>

**Before the first l**:
- "He" Between the first and second l: "" (empty string, as there’s no character between the two ls)

<br>

**After the second l**: "o! Wor"

- The resulting list is: ["He", "", "o!, Wor", "d"].

<br>

**Output:**

In [3]:
words: str = my_string.split("l") # Splitting using 'l' as the delimiter
print(words)

['He', '', 'o! Wor', 'd']


The argument of join() is the iterable, and the my_string acts as a separator that is placed between the elements of the iterable.

my_string = '`, `'

The variable my_string is assigned the value '`, `'. This string (`comma followed by a space`) will be used as the `separator` for joining elements of a list.

In [ ]:
# join the words back into a single string
my_string: str = ', '
joined_string: str = my_string.join(['Pakistan', 'USA', 'Canada', 'France', 'Japan'])
print(joined_string)  # Pakistan, USA, Canada, France, Japan

Pakistan, USA, Canada, France, Japan


In [ ]:
joined_string: str = my_string.join('Pakistan') # my_string works as a seprator for each character in the word 'Pakistan', because string is a sequence of caharacter

print(joined_string) # P, a, k, i, s, t, a, n

print('-'.join(['Apple', 'Banana', 'Cherry'])); # ; The line terminitor

P, a, k, i, s, t, a, n
Apple-Banana-Cherry


In [ ]:
my_string: str = "Hello, World! Hello, Pakistan"
# replace a substring
my_string = my_string.replace('Hello', 'Salam Walikum')
print(my_string)  # prints 'Salam Walikum, World! Salam Walikum, Pakistan'

Salam Walikum, World! Salam Walikum, Pakistan


In [ ]:
my_string: str = "Hello, World! Hello, Pakistan"
# find the index of a substring
starting_index = my_string.find('Hello') # Index value of the first occurance of the word 'Hello'
print("starting_index = ", starting_index)  # prints 0

starting_index =  0


In [ ]:
# Now lets find the second occurance of the word 'Hello'
# index value start from zero

starting_index2: str = starting_index + len("Hello") #len=5

print(my_string[starting_index2:] ) # after slicing ", World! Hello, Pakistan
print(my_string[starting_index2:].find("Hello")) # Starting index 9, because of slicing ", World! Hello, Pakistan"

, World! Hello, Pakistan
9


In [ ]:
print(my_string)
print(len(my_string)) #character count = 29

print('\n-----\n')

print("Substring to search    = ", "Hello")
print("Starting index         = ", len("Hello")) # 5
print("End index              = ", len(my_string))
print("Second Occurance index = ", my_string.index(("Hello"), len("Hello"), len(my_string))) # (substring, start, end) - print 14

#Uncomment to see ValueError: substring not found
#print("Second Occurance index = ", my_string.index(("Hello7"), len("Hello"), len(my_string))) # (substring, start, end)

Hello, World! Hello, Pakistan
29

-----

Substring to search    =  Hello
Starting index         =  5
End index              =  29
Second Occurance index =  14


In [ ]:
print(my_string.find("Hello7")) # -1 for not found

-1


## Conclusion
-  index() → Finds substring, raises error if not found.
-  find() → Finds substring, returns -1 if not found (safer).
-  Can specify start and end positions for more control.
-  Use a loop if you need all occurrences of a substring.

In [ ]:
my_string = "Hello, World! Hello, Pakistan"
# count the occurrences of a substring
count = my_string.count('Hello')
print("my_string.count('Hello') = ", count)  # prints 2

# count the occurrences of a substring
count = my_string.count('P')
print("my_string.count('P')     = ", count)  # prints 1

# count the occurrences of a substring
count = my_string.count('o')
print("my_string.count('o')     = ", count)  # prints 2

# count the occurrences of a substring
count = my_string.count('hello') # case sensitive
print("my_string.count('hello') = ", count)  # prints 0

my_string.count('Hello') =  2
my_string.count('P')     =  1
my_string.count('o')     =  3
my_string.count('hello') =  0


**String Formatting**
-------------------

Python provides several ways to format strings:

1. **`%` Operator**: `my_string = 'Hello, %s!' % 'World'`
  * In Python, `%s`, `%d`, `%c`, `%f` are placeholders for values in a string. They are used with the `%` operator to insert values into a string.
  * **Note**: The `%` operator is an older way of formatting strings in Python. The newer and more recommended way is to use the `str.format()` method or f-strings (introduced in Python 3.6). For example:
2. **`str.format()`**: `my_string = 'Hello, {}!'.format('World')`
3. **f-Strings**: `my_string = f'Hello, {"World"}!'`


<br>

| Placeholder | Meaning                              | Example                                |
|-------------|--------------------------------------|----------------------------------------|
| %s          | String                               | "Hello, %s" % "Alice" → "Hello, Alice" |
| %d          | Integer (Decimal)                    | "Age: %d" % 25 → "Age: 25"             |
| %c          | Character                            | "Letter: %c" % 'A' → "Letter: A"       |
| %f          | Floating-point                       | "Pi: %f" % 3.14159 → "Pi: 3.141590"    |
| %.nf        | Floating-point with n decimal places | "%.2f" % 3.14159 → "3.14"              |


## **% Operator**

In [ ]:
name: str = 'John'
age: int = 20
first_letter: str = name[0]
my_weight: float = 70.532000 # 70.536000

#uncomment to see type
#print(type((name, first_letter, age, my_weight)))

# using % operator
my_string: str = '''My name is %s, first letter of my name is \'%c\', I am %d years old and my weight id %f Kg.''' % (name, first_letter, age, my_weight)
print(my_string)

my_string = '''My name is %s, first letter of my name is \'%c\', I am %d years old and my weight id %.2f Kg.''' % (name, first_letter, age, my_weight) # Dont forget period %.2f
print(my_string)

My name is John, first letter of my name is 'J', I am 20 years old and my weight id 70.532000 Kg.
My name is John, first letter of my name is 'J', I am 20 years old and my weight id 70.53 Kg.


## **String Formatting in Python – Does Order Matter?**

**Yes**, order matters when using string formatting in Python, especially with the % operator, .format(), and f-strings (f""). Let's break it down:

In [ ]:
# Error: TypeError: %d format: a real number is required, not str
my_string: str = '''My name is %s, first letter of my name is \'%c\', I am %d years old and my weight id %f Kg.''' % (my_weight, age, name, first_letter)

TypeError: %d format: a real number is required, not str

## **str.format()**

In [ ]:
# using str.format()
my_string: str = 'My name is {} and I am {} years old.'.format(age, name) #order matters
print("line 1: ",my_string)

my_string: str = 'My name is {1} and I am {0} years old.'.format(age, name) #use indexing
print("line 2: ",my_string)

line 1:  My name is 20 and I am John years old.
line 2:  My name is John and I am 20 years old.


## **F-String**

In [ ]:
# using f-strings
my_string: str = f'My name is {name} and I am {age} years old.' #Using Named Placeholders (Best for Readability)
print("line 3: ",my_string)

my_string: str = fr'My \name is {name} and I am {age}\n \t years \old.' #At the same time it could be f and r as well
print("line 4: ",my_string)

line 3:  My name is John and I am 20 years old.
line 4:  My \name is John and I am 20\n \t years \old.



---

### 🔤 **String Literal Pool ka matlab kya hai?**

Python mein jab aap koi **string** likhtay ho (jaise `"hello"`), to Python yeh check karta hai:

> **"Kya yeh string pehle se memory mein maujood hai?"**

Agar haan, to **nayi string create nahi karta**, balke **pehli wali string ka reference** (link) de deta hai. Is process ko kehtay hain: **String Interning**.

---

### 🔧 **Kaise kaam karta hai?**

Jab Python koi string dekhta hai to yeh 3 steps follow karta hai:

---

#### **1. Interning Check**  
Sabse pehle yeh dekhta hai:
> Kya yeh string pehle se intern (yaani stored) hai?

Agar haan:  
✅ to **wahi purani string ka link de deta hai**

---

#### **2. String Literal Pool Check**  
Agar interned string nahi milti, to:
> Python ek **pool** (yaani ek chhoti si memory list) check karta hai jisme recent strings hoti hain.

Agar wahan mil gayi:  
✅ to **uska reference de deta hai**

---

#### **3. New String Banani**  
Agar dono jaga string nahi milti:  
❌ to Python **nayi string banata hai** aur memory mein daal deta hai.

---

### 🎁 **Fayda kya hai? (Benefits)**

#### ✅ Memory bachti hai  
Ek hi string bar bar likhne par nayi memory nahi leti → **memory save hoti hai**

#### ⚡ Fast hota hai  
Purani string ka link lena zyada fast hota hai → **performance improve hoti hai**

#### 💡 Comparisons easy  
Jab 2 same strings compare karo, Python easily check kar leta hai (kyunke dono ka reference same hota hai)

---

### 🧠 Simple Example:
```python
a = "hello"
b = "hello"
print(a is b)  # Output: True
```
Dono same string likhi hai, to Python `a` aur `b` ko **ek hi jagah se point karwata hai** → isiliye `a is b` = `True`

---



In [ ]:
a: str = "hello"
b: str = "hello"

#Both variable a and b have same id.
print(id(a))
print(id(b))

140381252799536
140381252799536


In [ ]:
c: str = a +"" # Nothing happen because we are appending a empty string.
print(id(c))

140381252799536


In [ ]:
c: str = a +" " # A new string object is created in the pool.
print(id(c))

140380450274224


## **Interning in Python**

### **1.  What is Interning?**

Interning is a process where Python stores only one copy of certain strings in memory and reuses them whenever the same string is used again. This is done to save memory and speed up comparisons.

### **2. Which Strings are Interned?**

* **Short strings**: Usually strings that are **`20 characters or fewer`**.
* **Identifiers**: Strings that look like variable names (e.g., "x", "my_var").
* **Explicit interning**: You can manually intern a string using the sys.intern() function.

### **3. Example of Manual Interning:**

In [ ]:
import sys
a = sys.intern("hello world!")
b = sys.intern("hello world!")
print(a is b)  # True (manually interned)

True


## **String Literal Pool in Python**

### **1.  What is the String Literal Pool?**

The string literal pool is a cache of string literals that appear in the code. When Python encounters a string literal (e.g., "hello"), it checks the pool to see if the string already exists. If it does, Python reuses the existing string instead of creating a new one.

### **2. How Does It Work?**

* **The pool is used for string literals explicitly written in the code** (e.g., "hello", "world").
* **It does not apply to dynamically created strings** (e.g., "hello" + "world").

### **3. Example of String Literal Pool:**

In [ ]:
a = "hello"
b = "hello"
print(a is b)  # True (both refer to the same string in the pool)

True




---

## 🧵 **1. String Pool kya hota hai?**

> **String pool ek memory area hota hai jahan Python kuch special strings ko store karta hai.**

- Jab aap koi string literal likhtay ho, jaise `"hello"`, Python usko string pool mein store karta hai.
- Taake agar future mein dubara `"hello"` likha jaye, to **nayi string banani na pare**.
- Yeh automatic hota hai. **Aapko kuch nahi karna parta.**

🧠 **Example:**
```python
a = "hello"
b = "hello"
print(a is b)  # True
```
Yahan `a` aur `b` same string pool wali jagah ko point kar rahe hain.

---

## 🔧 **2. `intern()` function kya karta hai?**

> `intern()` ek built-in function hai jo **aapki banai hui string ko manually string pool mein daal deta hai.**

Agar aap kisi string ko concatenate karke ya kisi aur tareeke se banate ho, to wo **automatic string pool mein nahi jaati**.

Aise mein, agar aap chaho to `sys.intern()` use karke manually usay string pool mein daal saktay ho.

🧠 **Example:**
```python
import sys

a = "hello world"
b = "hello" + " world"
c = sys.intern(a)

print(a is b)  # True hoga agar Python ne optimize kiya ho
print(a is c)  # Yeh hamesha True hoga, kyunke intern() se force kiya
```

---

## 📊 **Fark Summary Table:**

| Feature               | String Pool                             | `intern()` Function                   |
|-----------------------|------------------------------------------|----------------------------------------|
| **Definition**         | Memory jahan Python common strings rakhta hai | Function to forcefully put a string into that pool |
| **Kaam hota hai?**     | Automatic                                | Manually karna padta hai (`sys.intern()`) |
| **Kab useful?**        | Jab string literals ho                  | Jab runtime mein string ban rahi ho    |
| **Control kis ke paas?** | Python interpreter                     | Aap ke paas (`intern()` use kar ke)     |

---



## **String Interning vs String Pool:**

String interning and string pool are two concepts that are often confused with each other, but they are not exactly the same thing.

### **String Interning:**

String interning is a process where the **`compiler or interpreter`** checks if a string already exists in memory before creating a new one. If the string already exists, the compiler or interpreter returns a reference to the existing string instead of creating a new one. This process is also known as "string caching" or "string sharing".


### **String Pool:**

A string pool, on the other hand, is a memory space where all the strings in a program are stored. When a string is created, it is added to the string pool. The string pool is used to manage the memory for string objects.

### **Here are the key differences between string interning and string pool:**

1. **Scope:** String interning is a process that can be applied to **`any type of object`**, **`not just strings`**. String pool, on the other hand, is a memory space specifically designed for strings.

2. **Purpose:** The purpose of string interning is to optimize memory usage by avoiding duplicate objects. The purpose of a string pool is to manage the memory for string objects.

3. **Implementation:** String interning can be implemented using a hash table or a dictionary, where the keys are the strings and the values are the corresponding string objects. A string pool, on the other hand, is typically implemented as a contiguous block of memory where all the strings are stored.

4. **Behavior:** String interning can be enabled or disabled depending on the language or framework. A string pool, on the other hand, is typically always enabled for strings.

<br>

### **Process = Interning**
### **Memory Space = Pool of String Litrels**




---

### 🔹 **String Interning kya hota hai?**

**Interning** ka matlab hota hai:  
> "Ek hi cheez ko dobara dobara na banana — agar pehle se hai to wahi use karo."

Jab hum string interning karte hain, to agar koi **same string pehle se memory me maujood hai**, to **new copy banane ke bajaye** wohi purani string ka reference de diya jata hai.

Yani:  
```
agar "Hello" already hai → to new "Hello" nahi banega  
→ bas purana "Hello" ka reference use hoga
```

---

### 🔹 **String Pool kya hota hai?**

**String Pool** ek **special jagah (memory area)** hoti hai jahan **saari string literals** store hoti hain.

Jab tum koi string **literal** likhtay ho jaise `"apple"` ya `"hello"`:
> woh **automatically String Pool** me chali jati hai.

---

### 🔁 **Interning vs Pool ka Asaan Farq**

| Feature          | String Interning                        | String Pool                             |
|------------------|------------------------------------------|------------------------------------------|
| **Kaam kya karta hai?** | Duplicate string banne se rokta hai | Strings ko store karnay ki jagah hoti hai |
| **Kahan use hota hai?** | Har type ke object me ho sakta hai | Sirf strings ke saath hota hai           |
| **Kaise hota hai?** | Hash Table ya Dictionary se            | Ek special memory block me               |
| **Enable/Disable?** | Language per depend karta hai          | Strings ke liye usually always on        |

---

### 🧪 **Python Code Example**

```python
# Do strings banaye
a = "hello"
b = "hello"

print(a is b)  # True because both come from string pool

# Ab ek new string object banaye using str()
c = str("hello")

print(b is c)  # Still True in CPython (due to interning)

# Ab explicitly intern karte hain
import sys

x = sys.intern("world")
y = sys.intern("world")

print(x is y)  # True, dono interned hain

# Without interning
m = "world"
n = "".join(["w", "o", "r", "l", "d"])  # creates new string object

print(m is n)  # False, kyunke n interned nahi hai
```

---

### 🎯 **Ek Line me Summary**

📌 `String Pool` ek jagah hai jahan string literals store hote hain.  
📌 `String Interning` ek process hai jisse **duplicate strings** avoid ki jati hain.



## **When Are Strings Not Interned or Pooled?**

1.  **Long Strings**: Strings longer than **`20 characters`** are usually not interned automatically.

In [ ]:
a = "this is a very long string"
b = "this is a very long string"
print(a is b)  # False (not interned)
print("id(a)", id(a))
print("id(b)", id(b))

False
id(a) 132179098748208
id(b) 132178196687376


In [4]:
import sys

a = sys.intern("this is a very long string")
b = sys.intern("this is a very long string")

print(a is b)  # True (manually interned)
print("id(a)", id(a))
print("id(b)", id(b))

True
id(a) 2044050013472
id(b) 2044050013472


2. **Dynamically Created Strings**: Strings created at runtime (e.g., concatenation) are not interned or pooled.

In [ ]:
a = "hello"
b = "world"
c = a + b  # Dynamically created string
d = "helloworld"
print("c is d = ", c is d)  # False (not interned or pooled)
print(c, " - id(c)", id(c))
print(d, " - id(d)", id(d))

c is d =  False
helloworld  - id(c) 132178196390256
helloworld  - id(d) 132178748685680


In [ ]:
c1 = a + b
print("c is c1 = ",  c is c1)

print(c1, " - id(c1)", id(c1))
print(c,  " - id(c) ", id(c))

print("is contant same = ", c == c1)

c is c1 =  False
helloworld  - id(c1) 132178196306416
helloworld  - id(c)  132178196390256
is contant same =  True


In [ ]:
# prompt: print list of str functions using dir(), dont show function starting with"__"

# Get the list of string methods
string_methods: str = dir(str)

# Filter out methods starting with "__"
filtered_methods: str = [method for method in string_methods if not method.startswith("__")]

# Print the filtered list
filtered_methods

['capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',
 'zfill']

## **String Repetition**

String repetition in Python is a straightforward way to create a new string by concatenating multiple copies of an existing string. You achieve this using the asterisk (`*`) operator.

Here's a concise breakdown:

-   **The Basics:**
    
    -   You take a string and multiply it by an integer.
    -   The result is a new string containing the original string repeated the specified number of times.
-   **Example:**
    
    -   `"abc" * 3` results in `"abcabcabc"`.
    -   `"Python! " * 2` results in `"Python! Python! "`.
-   **Key Points:**
    
    -   The `*` operator requires an integer as the multiplier.
    -   Multiplying a string by zero results in an empty string.
    -   This is a very efficient way to create a longer string when you need repeating patterns.
-   **Use Cases:**
    
    -   Creating visual separators (e.g., `"-" * 20`).
    -   Generating repetitive patterns for text-based displays.
    -   Quickly building strings with repeated elements.



---

### 🔹 **String Repetition kya hoti hai?**

**String repetition** ka matlab hota hai:
> "Ek string ko baar baar repeat karna (dobara likhna) without copy-pasting."

Python mein hum **`*` (asterisk)** ka use kar ke ye kaam karte hain.

---

### 🧠 **Kaise kaam karta hai?**

Tum likhtay ho:  
```python
"string" * number
```

Yani:
- `"hello" * 3` → `"hellohellohello"`
- `"A! " * 5` → `"A! A! A! A! A! "`

---

### 📌 **Important Baatain**

| Rule | Asaan Samajh |
|------|--------------|
| `*` ka right side me **number hona chahiye** | `"hi" * 3` — ✅ thik hai |
| Agar **0 se multiply karoge**, to khaali string milegi | `"hi" * 0` → `""` |
| Integers ke ilawa kuch aur diya to error milega | `"hi" * "2"` — ❌ wrong |

---

### ✅ **Use Cases / Kahan kaam aata hai?**

- Visual line banana:  
  `"=" * 30` → `==============================`

- Design pattern banana:  
  `"abc-" * 4` → `"abc-abc-abc-abc-"`

- Empty string banana:  
  `"test" * 0` → `""` (useful in condition check)

---

### 💻 **Simple Python Example**

```python
print("Hello " * 3)
# Output: Hello Hello Hello 

print("-" * 20)
# Output: --------------------

stars = "* " * 5
print(stars)
# Output: * * * * * 

pattern = "abc-" * 2
print(pattern)
# Output: abc-abc-
```

---

### 🪄 **Ek Line Summary**

📌 `string * number` se tum ek string ko repeat kar sakte ho — bohat simple aur fast tareeqa!





---



In [ ]:
# String repetition examples

base_string = "Hello"
repetition_count = 3

repeated_string = base_string * repetition_count

print(f"Original string: {base_string}")
print(f"Repeated string: {repeated_string}")

# Using repetition for visual separators
separator = "-" * 30
print(separator)

# Repeating with different strings and counts
pattern = "* "
repeated_pattern = pattern * 5
print(repeated_pattern)

# Repeating zero times
empty_string = "Test" * 0
print(f"Empty string: '{empty_string}'")

#using string repetition in a loop.
for i in range(1,6):
  print("*"*i)

Original string: Hello
Repeated string: HelloHelloHello
------------------------------
* * * * * 
Empty string: ''
*
**
***
****
*****


## **Explanation:**

1.  **Basic Repetition:**
    -   We define a `base_string` and a `repetition_count`.
    -   The `*` operator multiplies the string by the count, creating `repeated_string`.
    -   f-strings are used to print the results.
2.  **Visual Separators:**
    -   We create a separator line using `"-" * 30`.
3.  **Different Patterns:**
    -   We demostrate creating a different repeating pattern.
4.  **Zero Repetition:**
    -   Demonstrates that multiplying by zero results in an empty string.
5.  **Loop Example**
    -   A for loop is used to print an increasing number of asterisks on each line, creating a triangle like pattern.


### **Comparison Operators for Strings in Python**

In Python, comparison operators work with **strings** based on **lexicographical order** (dictionary order), which follows the **Unicode values** of characters.

### **Examples of String Comparisons**

In [ ]:
# Define two strings
str1 = "apple"
str2 = "banana"

print("str1 == str2 = ",str1 == str2)  # False (because "apple" is not equal to "banana")
print("str1 != str2 = ",str1 != str2)  # True  (because "apple" is different from "banana")
print("str1 > str2  = ", str1 > str2)   # False (because "apple" comes before "banana" in dictionary order)
print("str1 < str2  = ", str1 < str2)   # True  (because "apple" comes before "banana")


str1 == str2 =  False
str1 != str2 =  True
str1 > str2  =  False
str1 < str2  =  True



---

### 🔡 **Strings kya hoti hain?**

String matlab: **Alphabets ka group**  
Jaise: `"apple"`, `"banana"`, `"car"`, `"zebra"`

---

### 📖 **Alphabetical Order kya hota hai?**

Socho ke tumhara paas dictionary (qaamoos) hai.  
Usme words kis order mein hote hain?

```text
apple  
banana  
cat  
dog  
```

➡️ Pehle **"a" wale** words  
➡️ Phir **"b" wale**  
➡️ Phir **"c" wale**

Yani:

🟢 `"apple"` < `"banana"`  
🔴 `"zebra"` > `"apple"`

---

### 🧠 **Code Example Samjho Jaise Real Life**

```python
str1 = "apple"
str2 = "banana"
```

Ye dono **words** hain jaise dictionary mein hote hain.

---

### 💬 Comparison One by One:

#### 1. `str1 == str2`

```python
"apple" == "banana"
```

🤔 Kya dono same hain?

❌ **Nahi!** Apple aur banana alag alag cheezein hain.  
✅ Output: `False`

---

#### 2. `str1 != str2`

```python
"apple" != "banana"
```

🤔 Kya dono alag hain?

✅ **Haan!** Apple ≠ Banana  
➡️ Output: `True`

---

#### 3. `str1 > str2`

```python
"apple" > "banana"
```

🤔 Kya "apple" dictionary mein "banana" ke baad aata hai?

❌ **Nahi!** Apple to pehle aata hai  
➡️ Output: `False`

---

#### 4. `str1 < str2`

```python
"apple" < "banana"
```

🤔 Kya "apple" pehle aata hai "banana" se?

✅ **Haan!** Apple pehle aata hai  
➡️ Output: `True`

---

### 🎯 Super Simple Summary:

| Comparison       | Matlab kya hai?               | Result        |
|------------------|-------------------------------|---------------|
| `"apple" == "banana"` | Kya dono same hain?              | ❌ `False` |
| `"apple" != "banana"` | Kya dono alag hain?              | ✅ `True`  |
| `"apple" > "banana"`  | Kya apple dictionary mein baad aata hai? | ❌ `False` |
| `"apple" < "banana"`  | Kya apple dictionary mein pehle aata hai? | ✅ `True`  |

---




### **How String Comparison Works?**

-   **Python compares strings character by character** using their **Unicode values**.
    
-   `'a' < 'b'` because the Unicode value of `'a'` (97) is smaller than that of `'b'` (98).
    
-   `'apple' < 'banana'` → `'a' < 'b'`, so Python stops there.
    

### **Comparing Strings with Different Cases**

In [ ]:
print('"Apple" > "apple"   = ', "Apple" > "apple")  # False ('A' has a smaller Unicode value than 'a')
print('"apple" == "APPLE") = ', "apple" == "APPLE") # False (case-sensitive)
print('"abc" < "abd"       = ', "abc" < "abd")      # True ('c' < 'd')


"Apple" > "apple"   =  False
"apple" == "APPLE") =  False
"abc" < "abd"       =  True


## **Using Comparison Operators in an If Statement**

In [ ]:
word = "mango"

if word > "apple":
    print(f"{word} comes after apple in alphabetical order")
else:
    print(f"{word} comes before apple in alphabetical order")


mango comes after apple in alphabetical order




---

## 🔄 **1. Type Casting (General)**

🧠 **Type casting** ka matlab hota hai:

> **Ek data type ko doosray data type mein convert karna.**

### ✅ Examples:

```python
x = 5        # integer
y = float(x) # ab y ban gaya 5.0 (float)

a = "123"
b = int(a)   # ab b ban gaya 123 (integer)
```

📌 Yani:
- `int()`: kisi cheez ko number banata hai
- `float()`: point waala number banata hai
- `str()`: kisi bhi cheez ko string banata hai

---

## 📝 **2. String Casting (Specific Type of Casting)**

📌 **String casting** type casting ka ek part hai — jahan **kisi bhi data ko string mein convert** kiya jata hai.

> Kisi number, float, ya boolean ko `str()` function se string bana dete hain.

### ✅ Examples:

```python
num = 99
s = str(num)  # ab s ban gaya "99" (ye ab text hai, number nahi)
```

---

## 📊 **Farak Table (Simple Words)**

| Feature              | Type Casting                      | String Casting                    |
|----------------------|-----------------------------------|-----------------------------------|
| Matlab kya hai?      | Kisi bhi type ko doosray type mein badalna | Kisi bhi value ko **string** mein badalna |
| Example              | `int("123")`, `float(10)`         | `str(123)`, `str(True)`           |
| Scope                | Har data type pe lagta hai        | Sirf string banane ke liye hota hai |
| Function             | `int()`, `float()`, `bool()`, `str()` | `str()`                            |

---

## 💡 Real-Life Example:

Socho tumhara paas number hai:

```python
age = 18
```

Ab agar tum ye likho:

```python
print("Your age is " + age)   # ❌ Error aayega (string + number nahi hota)
```

To ye galat hai!

**Sahi tarika (string casting):**

```python
print("Your age is " + str(age))  # ✅ "Your age is 18"
```

---

## 🔚 Easy Summary:

🟢 **Type Casting** = Har type mein convert karna (int, float, str, bool)  
🟢 **String Casting** = Sirf string banane ke liye use hota hai (`str()`)

---



#**Comprehensive Guide to Type Casting in Python** 🚀

## **What is Type Casting?**

Type casting (or type conversion) is the process of converting one data type into another. Python supports two types of type casting:

- **`Implicit`** Type Casting – Done `automatically` by Python.
- **`Explicit`** Type Casting – Done `manually using built-in functions`.

##1️⃣ **Implicit Type Casting (Automatic Conversion)**

Python automatically converts one data type to another `when no data loss occurs.`

Example 1: Converting int to float (`Safe Conversion`)

In [ ]:
num_int: int = 10
num_float = num_int + 5.5  # int + float = float. skipped type hint to see what data type is assigned at runtime
print(num_float, type(num_float))

15.5 <class 'float'>


### **Example 2: Converting int to complex**

✅ Python automatically promotes int to complex, as complex numbers `cannot be downgraded`.

In [ ]:
num_int: int = 7
num_complex: complex = num_int + 3j  # int + complex → complex
print(num_complex, type(num_complex))

# num_int automatically promotes int to complex type
num_int = num_complex
print(num_int, type(num_int))

(7+3j) <class 'complex'>
(7+3j) <class 'complex'>


### **Example 3: No Implicit Conversion Between str and int**

❌ Error: Python does NOT implicitly convert str to int. We must convert explicitly.



In [ ]:
num_str = "100"
num_int = 5

print(num_str + num_int)  # ❌ TypeError

TypeError: can only concatenate str (not "int") to str

In [ ]:
print(int(num_str) + num_int) # ✅  casting str to int, arthmitic operation perfomed
print(num_str + str(num_int)) # ✅  casting int to str, string concatenation performed

105
1005


## **2️⃣ Explicit Type Casting (Manual Conversion)**

Python provides several built-in functions for type conversion:

<br>

| function   | type           |   |
|------------|----------------|---|
| int(x)     | Integer        |   |
| float(x)   | Float          |   |
| complex(x) | Complex number |   |
| str(x)     | String         |   |
| bool(x)    | Boolean        |   |
| list(x)    | List           |   |
| tuple(x)   | Tuple          |   |
| set(x)     | Set            |   |
| dict(x)    | Dictionary     |   |
|            |                |   |

### **Integer Conversion (int())**

Used to convert float, string (`if valid number`), and boolean values to integers.

✅ Example 1: Float → Integer (`Removes Decimal Part`)


🔹 Note: int() `truncates` (removes) the decimal part, it does not round.

In [ ]:
num_float: float = 9.8
num_int = int(num_float) # skipped type hint to see what data type is assigned at runtime
print(num_int, type(num_int))

b: bool = True
print("int(b) = ", int(b))

9 <class 'int'>
int(b) =  1


✅ Example 2: String → Integer (Only Valid Numbers)

❌ Invalid Case: "123abc" cannot be converted to int(). It will raise a ValueError.

In [ ]:
num_str: str = "123"
#num_str = "123abc" #uncomment to see error

num_int = int(num_str) # skipped type hint to see what data type is assigned at runtime
print(num_int, type(num_int))

123 <class 'int'>


✅ Example 3: Boolean → Integer

In [ ]:
print(int(True))   # Output: 1
print(int(False))  # Output: 0

1
0


## **Float Conversion (float())**

Converts integers, strings, and booleans to floating-point numbers.

✅ Example 1: Integer → Float

In [ ]:
num_int: int = 5
num_float = float(num_int) # skipped type hint to see what data type is assigned at runtime
print(num_float, type(num_float))

5.0 <class 'float'>


✅ Example 2: String → Float

In [ ]:
num_str: str = "3.14"
num_float = float(num_str) # skipped type hint to see what data type is assigned at runtime
print(num_float, type(num_float))

3.14 <class 'float'>


## **String Conversion (str())**

Converts numbers, lists, tuples, dictionaries, and booleans to strings.

✅ Example 1: Number → String

In [ ]:
num: int = 100
num_str = str(num) # skipped type hint to see what data type is assigned at runtime
print(num_str, type(num_str))

100 <class 'str'>


✅ Example 2: Boolean → String

In [ ]:
print(str(True))   # Output: "True"
print(str(False))  # Output: "False"

True
False


## **Boolean Conversion (bool())**

Converts other data types to boolean (True or False).

✅ Example:

In [ ]:
print("bool(1)       = ", bool(1))       # True
print("bool(0)       = ", bool(0))       # False
print("bool(-10)     = ", bool(-10))     # True (Non-zero numbers are True)
print('bool("")      = ', bool(""))      # False (Empty string)
print('bool("Hello") = ', bool("Hello")) # True (Non-empty string)
print("bool([])      = ", bool([]))      # False (Empty list)
print("bool([1, 2])  = ", bool([1, 2]))  # True (Non-empty list)

bool(1)       =  True
bool(0)       =  False
bool(-10)     =  True
bool("")      =  False
bool("Hello") =  True
bool([])      =  False
bool([1, 2])  =  True


## **List, Tuple & Set Conversions**

Python allows converting between lists, tuples, and sets.

✅ Example 1: Tuple → List

In [ ]:
tup: tuple = (1, 2.7, 3, 'OB')
lst = list(tup) # skipped type hint to see what data type is assigned at runtime
print(lst, type(lst))

[1, 2.7, 3, 'OB'] <class 'list'>


✅ Example 3: List → Set (Removes Duplicates)

In [ ]:
lst: list = [1, 2, 2, 3, 4, 4, 5, "Agentic AI"]
s = set(lst)      # skipped type hint to see what data type is assigned at runtime
print(s, type(s))

{1, 2, 3, 4, 5, 'Agentic AI'} <class 'set'>


## **Dictionary Conversion (dict())**

You can convert lists of key-value pairs into a dictionary.

✅ Example: List of Tuples → Dictionary

In [ ]:
lst: list = [("name", "Alice"), ("age", 25)]
d = dict(lst)       # skipped type hint to see what data type is assigned at runtime
print(d, type(d))

{'name': 'Alice', 'age': 25} <class 'dict'>


## **Complex Number Conversion (complex())**

Used to convert numbers into complex numbers (a + bj).

✅ Example:

In [ ]:
num: int = 5
comp = complex(num)   # skipped type hint to see what data type is assigned at runtime
print(comp, type(comp))  # Output: (5+0j) <class 'complex'>

(5+0j) <class 'complex'>


In [ ]:
num = str(comp)     # skipped type hint to see what data type is assigned at runtime
num

'(5+0j)'

In [ ]:
# TypeError: int() argument must be a string, a bytes-like object or a real number, not 'complex'
num = int(comp)
num

TypeError: int() argument must be a string, a bytes-like object or a real number, not 'complex'

## **Conclusion**

✅ Implicit Conversion (Automatic, No Data Loss)

✅ Explicit Conversion (int(), float(), str(), etc.) for manual control

✅ Be careful when converting strings (avoid invalid format errors)


---

### ✅ **Python Type Conversion Table (Implicit vs Explicit)**

| 🔢 From Type | 🔄 To Type | ⚙️ Conversion Type | ✅ Function / Example         |
|-------------|------------|--------------------|-------------------------------|
| `int`       | `float`    | Implicit           | `x = 5 + 2.0` → `7.0`         |
| `int`       | `complex`  | Implicit           | `complex(5)` → `(5+0j)`       |
| `bool`      | `int`      | Implicit           | `int(True)` → `1`            |
| `int`       | `str`      | Explicit           | `str(123)` → `"123"`         |
| `str`       | `int`      | Explicit           | `int("456")` → `456`         |
| `float`     | `int`      | Explicit           | `int(5.9)` → `5`             |
| `str`       | `float`    | Explicit           | `float("3.14")` → `3.14`     |
| `list`      | `set`      | Explicit           | `set([1, 2, 2])` → `{1, 2}`   |
| `tuple`     | `list`     | Explicit           | `list((1, 2))` → `[1, 2]`     |
| `str`       | `list`     | Explicit           | `list("abc")` → `['a','b','c']`|
| `str`       | `tuple`    | Explicit           | `tuple("ab")` → `('a','b')`  |
| `int`       | `bool`     | Explicit           | `bool(0)` → `False`          |
| `float`     | `str`      | Explicit           | `str(2.5)` → `"2.5"`         |

---





---

### **String Basics**
1. What will be the output of `print('Hello\\nWorld!')`?
   a) Hello\\nWorld!  
   b) Hello World!  
   c) Hello  
      World!  
   d) Error  
   **Answer:** c) Hello (newline) World!  
   **Explanation:** `\n` is an escape sequence for a newline character.

2. Which of these creates a multi-line string in Python?
   a) `'''Hello\\nWorld'''`  
   b) `'Hello World'`  
   c) `r'Hello\\nWorld'`  
   d) `'Hello' + 'World'`  
   **Answer:** a) `'''Hello\nWorld'''`  
   **Explanation:** Triple quotes allow multi-line strings, and `\n` adds a newline.

3. What does `r'Hello\\tWorld'` print?  
   a) Hello    World  
   b) Hello\\tWorld  
   c) Hello World  
   d) Error  
   **Answer:** b) Hello\\tWorld  
   **Explanation:** Raw strings (`r`) treat backslashes as literal characters.

---

### **String Operations**
4. What is the output of `'Hello'[1:4]`?  
   a) 'ell'  
   b) 'Hel'  
   c) 'ello'  
   d) 'llo'  
   **Answer:** a) 'ell'  
   **Explanation:** Slicing from index 1 to 3 (4 is exclusive).

5. What does `len(' Hello ')` return?  
   a) 5  
   b) 6  
   c) 7  
   d) 8  
   **Answer:** c) 7  
   **Explanation:** Spaces are counted as characters.

6. What is the result of `'Hello'.upper().lower()`?  
   a) 'HELLO'  
   b) 'hello'  
   c) 'Hello'  
   d) Error  
   **Answer:** b) 'hello'  
   **Explanation:** `upper()` converts to uppercase, then `lower()` converts back to lowercase.

---

### **String Methods**
7. What does `'Hello,World'.split(',')` return?  
   a) `['Hello', 'World']`  
   b) `['Hello,World']`  
   c) `['H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd']`  
   d) Error  
   **Answer:** a) `['Hello', 'World']`  
   **Explanation:** `split(',')` splits the string at commas.

8. What is the output of `'-'.join(['A', 'B', 'C'])`?  
   a) 'A-B-C'  
   b) 'A B C'  
   c) 'ABC'  
   d) Error  
   **Answer:** a) 'A-B-C'  
   **Explanation:** `join()` concatenates list elements with the specified separator.

9. What does `'Hello'.replace('l', 'x', 1)` return?  
   a) 'Hexlo'  
   b) 'Hexxo'  
   c) 'He llo'  
   d) 'Hello'  
   **Answer:** a) 'Hexlo'  
   **Explanation:** Only the first occurrence of 'l' is replaced due to the count parameter `1`.

---

### **String Formatting**
10. What is the output of `f'{\"Hello\"}'`?  
    a) 'Hello'  
    b) '{Hello}'  
    c) Error  
    d) 'fHello'  
    **Answer:** a) 'Hello'  
    **Explanation:** f-strings evaluate expressions inside curly braces.

11. Which line correctly formats a string with `%` operator?  
    a) `'Name: %d' % 'Alice'`  
    b) `'Name: %s' % 'Alice'`  
    c) `'Name: %f' % 'Alice'`  
    d) `'Name: %c' % 'Alice'`  
    **Answer:** b) `'Name: %s' % 'Alice'`  
    **Explanation:** `%s` is the placeholder for strings.

12. What does `'%.2f' % 3.14159` return?  
    a) '3.14'  
    b) '3.14159'  
    c) '3.1'  
    d) '3'  
    **Answer:** a) '3.14'  
    **Explanation:** `%.2f` rounds to 2 decimal places.

---

### **String Interning & Pool**
13. What is the output of `a = 'hello'; b = 'hello'; print(a is b)`?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** a) True  
    **Explanation:** Short strings are interned, so `a` and `b` point to the same object.

14. Which string will NOT be interned automatically?  
    a) `'x'`  
    b) `'hello'`  
    c) `'this is a very long string'`  
    d) `'python'`  
    **Answer:** c) `'this is a very long string'`  
    **Explanation:** Strings longer than 20 characters are not interned by default.

15. How can you manually intern a string?  
    a) `intern('hello')`  
    b) `sys.intern('hello')`  
    c) `str.intern('hello')`  
    d) `'hello'.intern()`  
    **Answer:** b) `sys.intern('hello')`  
    **Explanation:** The `sys.intern()` function forces interning.

---

### **Type Casting**
16. What does `int(3.9)` return?  
    a) 3  
    b) 4  
    c) 3.9  
    d) Error  
    **Answer:** a) 3  
    **Explanation:** `int()` truncates the decimal part.

17. What is the output of `bool('False')`?  
    a) False  
    b) True  
    c) Error  
    d) None  
    **Answer:** b) True  
    **Explanation:** Non-empty strings evaluate to `True`.

18. Which conversion will raise a `ValueError`?  
    a) `int('123')`  
    b) `float('3.14')`  
    c) `int('abc')`  
    d) `str(100)`  
    **Answer:** c) `int('abc')`  
    **Explanation:** `'abc'` cannot be converted to an integer.

19. What does `list('hello')` return?  
    a) `['h', 'e', 'l', 'l', 'o']`  
    b) `['hello']`  
    c) `'hello'`  
    d) Error  
    **Answer:** a) `['h', 'e', 'l', 'l', 'o']`  
    **Explanation:** Strings are iterable; `list()` converts each character to a list element.

20. What is the output of `set([1, 2, 2, 3])`?  
    a) `{1, 2, 2, 3}`  
    b) `{1, 2, 3}`  
    c) `[1, 2, 3]`  
    d) Error  
    **Answer:** b) `{1, 2, 3}`  
    **Explanation:** Sets remove duplicates.

---

### **Advanced String Questions**
21. What does `'Hello'.find('l')` return?  
    a) 2  
    b) 3  
    c) [2, 3]  
    d) -1  
    **Answer:** a) 2  
    **Explanation:** Returns the index of the first occurrence of 'l'.

22. What is the output of `'Hello'.index('x')`?  
    a) -1  
    b) None  
    c) Error  
    d) 0  
    **Answer:** c) Error  
    **Explanation:** `index()` raises a `ValueError` if the substring is not found.

23. What does `'Hello'.count('l')` return?  
    a) 1  
    b) 2  
    c) 0  
    d) Error  
    **Answer:** b) 2  
    **Explanation:** There are two 'l' characters in "Hello".

24. What is the result of `'Hello' + 1`?  
    a) 'Hello1'  
    b) 'Hellow'  
    c) Error  
    d) 'Hello '  
    **Answer:** c) Error  
    **Explanation:** Cannot concatenate string and integer directly.

25. What does `'Hello' * 3` return?  
    a) 'HelloHelloHello'  
    b) 'Hello3'  
    c) Error  
    d) 'Hello Hello Hello'  
    **Answer:** a) 'HelloHelloHello'  
    **Explanation:** String repetition concatenates the string `n` times.

---

### **Tricky Comparisons**
26. What is the output of `'apple' > 'banana'`?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** b) False  
    **Explanation:** 'a' (apple) has a lower Unicode value than 'b' (banana).

27. What does `'A' == 'a'` return?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** b) False  
    **Explanation:** String comparisons are case-sensitive.

28. What is the result of `'10' < '2'`?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** a) True  
    **Explanation:** Strings are compared lexicographically; '1' < '2'.

---

### **Type Casting Scenarios**
29. What does `float(True)` return?  
    a) 1.0  
    b) True  
    c) Error  
    d) None  
    **Answer:** a) 1.0  
    **Explanation:** `True` is converted to `1.0` in float.

30. What is the output of `str(3.14)[::-1]`?  
    a) '3.14'  
    b) '41.3'  
    c) Error  
    d) '31.4'  
    **Answer:** b) '41.3'  
    **Explanation:** `[::-1]` reverses the string `'3.14'`.

31. What does `bool(' ')` return?  
    a) False  
    b) True  
    c) Error  
    d) None  
    **Answer:** b) True  
    **Explanation:** A space is a non-empty string.

32. What is the result of `int('0b101', 2)`?  
    a) 5  
    b) 101  
    c) Error  
    d) 0  
    **Answer:** a) 5  
    **Explanation:** `int()` with base 2 converts binary to decimal.

---

### **String Pool & Interning**
33. What is the output of `a = 'hello'; b = 'hello'; print(id(a) == id(b))`?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** a) True  
    **Explanation:** Short strings are interned, so `a` and `b` share the same memory.

34. Which of these strings will NOT be interned automatically?  
    a) `'x'`  
    b) `'python'`  
    c) `'a' * 20`  
    d) `'a' * 21`  
    **Answer:** d) `'a' * 21`  
    **Explanation:** Strings longer than 20 characters are not interned by default.

---


### **Escape Sequences**
35. What does `print('Hello\\bWorld')` output?  
    a) HelloWorld  
    b) HellWorld  
    c) Hello\bWorld  
    d) Error  
    **Answer:** b) HellWorld  
    **Explanation:** `\b` is a backspace escape sequence.

36. What is the output of `print('\u0041')`?  
    a) 'A'  
    b) '\\u0041'  
    c) Error  
    d) None  
    **Answer:** a) 'A'  
    **Explanation:** `\u0041` is the Unicode for 'A'.

---

### **String Methods Deep Dive**
37. What does `'Hello'.isalpha()` return?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** a) True  
    **Explanation:** All characters in 'Hello' are alphabetic.

38. What is the output of `'123'.isdigit()`?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** a) True  
    **Explanation:** The string consists only of digits.

39. What does `'Hello World'.title()` return?  
    a) 'Hello world'  
    b) 'Hello World'  
    c) 'HELLO WORLD'  
    d) 'Hello World'  
    **Answer:** b) 'Hello World'  
    **Explanation:** `title()` capitalizes the first letter of each word.

---

### **Complex Type Casting**
40. What is the output of `complex(5)`?  
    a) `5`  
    b) `5 + 0j`  
    c) `5j`  
    d) Error  
    **Answer:** b) `5 + 0j`  
    **Explanation:** `complex()` converts integers to complex numbers.

41. What does `str(True) + str(False)` return?  
    a) 'TrueFalse'  
    b) '10'  
    c) Error  
    d) '1'  
    **Answer:** a) 'TrueFalse'  
    **Explanation:** Boolean values are converted to strings and concatenated.

---

### **String Slicing Tricks**
42. What is the output of `'Hello'[100:]`?  
    a) Error  
    b) 'Hello'  
    c) ''  
    d) 'o'  
    **Answer:** c) ''  
    **Explanation:** Out-of-bounds slicing returns an empty string.

43. What does `'Hello'[::-1]` return?  
    a) 'Hello'  
    b) 'olleH'  
    c) Error  
    d) 'H'  
    **Answer:** b) 'olleH'  
    **Explanation:** `[::-1]` reverses the string.

---

### **String Formatting Nuances**
44. What is the output of `'{} {}'.format(1, 2, 3)`?  
    a) '1 2'  
    b) '1 2 3'  
    c) Error  
    d) '3 2'  
    **Answer:** a) '1 2'  
    **Explanation:** Extra arguments are ignored in `format()`.

45. What does `f'{10 + 5}'` return?  
    a) '10 + 5'  
    b) '15'  
    c) Error  
    d) '105'  
    **Answer:** b) '15'  
    **Explanation:** f-strings evaluate expressions inside curly braces.

---

### **Advanced Comparisons**
46. What is the output of `'a' < 'A'`?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** b) False  
    **Explanation:** 'a' (97) has a higher Unicode value than 'A' (65).

47. What does `'10' == 10` return?  
    a) True  
    b) False  
    c) Error  
    d) None  
    **Answer:** b) False  
    **Explanation:** String and integer are not equal.

---

### **String Repetition**
48. What is the output of `'Hi' * 0`?  
    a) 'Hi'  
    b) ''  
    c) Error  
    d) '0'  
    **Answer:** b) ''  
    **Explanation:** Multiplying a string by 0 returns an empty string.

49. What does `'A' * 3 + 'B' * 2` return?  
    a) 'AAABB'  
    b) 'ABABAB'  
    c) 'A3B2'  
    d) Error  
    **Answer:** a) 'AAABB'  
    **Explanation:** String repetition is performed before concatenation.

---

### **Final Tricky Question**
50. What is the output of `print('Hello\\n'.strip())`?  
    a) 'Hello\\n'  
    b) 'Hello'  
    c) 'Hell'  
    d) Error  
    **Answer:** b) 'Hello'  
    **Explanation:** `strip()` removes leading/trailing whitespace, including `\n`.

---

